osu.Game.Tests/NonVisual/Filtering/FilterQueryParserTest.cs

```diff
@@ -120,24 +120,54 @@ public void TestApplyBPMQueries()
-            new object[] { "70", TimeSpan.FromSeconds(70), TimeSpan.FromSeconds(1) },
+            new object[] { "23s", TimeSpan.FromSeconds(23), TimeSpan.FromSeconds(1), true },
+            new object[] { "9m", TimeSpan.FromMinutes(9), TimeSpan.FromMinutes(1), true },
+            new object[] { "0.25h", TimeSpan.FromHours(0.25), TimeSpan.FromHours(1), true },
+            new object[] { "70", TimeSpan.FromSeconds(70), TimeSpan.FromSeconds(1), true },
+            new object[] { "7m27s", TimeSpan.FromSeconds(447), TimeSpan.FromSeconds(1), true },
+            new object[] { "7:27", TimeSpan.FromSeconds(447), TimeSpan.FromSeconds(1), true },
+            new object[] { "1h2m3s", TimeSpan.FromSeconds(3723), TimeSpan.FromSeconds(1), true },
+            new object[] { "1h2m3.5s", TimeSpan.FromSeconds(3723.5), TimeSpan.FromSeconds(1), true },
+            new object[] { "1:2:3", TimeSpan.FromSeconds(3723), TimeSpan.FromSeconds(1), true },
+            new object[] { "1:02:03", TimeSpan.FromSeconds(3723), TimeSpan.FromSeconds(1), true },
+            new object[] { "6", TimeSpan.FromSeconds(6), TimeSpan.FromSeconds(1), true },
+            new object[] { "6.5", TimeSpan.FromSeconds(6.5), TimeSpan.FromSeconds(1), true },
+            new object[] { "6.5s", TimeSpan.FromSeconds(6.5), TimeSpan.FromSeconds(1), true },
+            new object[] { "6.5m", TimeSpan.FromMinutes(6.5), TimeSpan.FromMinutes(1), true },
+            new object[] { "6h5m", TimeSpan.FromMinutes(365), TimeSpan.FromMinutes(1), true },
+            new object[] { "65m", TimeSpan.FromMinutes(65), TimeSpan.FromMinutes(1), true },
+            new object[] { "90s", TimeSpan.FromSeconds(90), TimeSpan.FromSeconds(1), true },
+            new object[] { "80m20s", TimeSpan.FromSeconds(4820), TimeSpan.FromSeconds(1), true },
+            new object[] { "1h20s", TimeSpan.FromSeconds(3620), TimeSpan.FromSeconds(1), true },
+            new object[] { "7.5m27s", new TimeSpan(), new TimeSpan(), false },
+            new object[] { "7m27", new TimeSpan(), new TimeSpan(), false },
+            new object[] { "7m7m7m", new TimeSpan(), new TimeSpan(), false },
+            new object[] { "7m70s", new TimeSpan(), new TimeSpan(), false },
+            new object[] { "5s6m", new TimeSpan(), new TimeSpan(), false },
+            new object[] { "0:", new TimeSpan(), new TimeSpan(), false },
+            new object[] { ":0", new TimeSpan(), new TimeSpan(), false },
+            new object[] { "0:3:", new TimeSpan(), new TimeSpan(), false },
+            new object[] { "3:15.5", new TimeSpan(), new TimeSpan(), false },
         };
 
         [Test]
         [TestCaseSource(nameof(length_query_examples))]
-        public void TestApplyLengthQueries(string lengthQuery, TimeSpan expectedLength, TimeSpan scale)
+        public void TestApplyLengthQueries(string lengthQuery, TimeSpan expectedLength, TimeSpan scale, bool isValid)
```

Comment:
I don't generally much like when a single test has conditional branches. It complicates understanding, and also - as can be seen from `length_query_examples` - you have to pass garbage data to the negative test cases just to fit the method signature. I'd split the positive cases and the negative cases into separate test cases, like so:

```csharp
[Test]
[TestCaseSource(nameof(correct_length_query_examples))]
public void TestApplyLengthQueries(string lengthQuery, TimeSpan expectedLength, TimeSpan scale)
{
    // ...
}

[Test]
[TestCaseSource(nameof(incorrect_length_query_examples))]
public void TestInvalidLengthQueries(string query)
{
    // ...
}
```